src/tree.rs

```diff
@@ -1,15 +1,298 @@
+use crate::{
+    constants::{
+        util::gen_const_array, L_ORCHARD_MERKLE, MERKLE_CRH_PERSONALIZATION, MERKLE_DEPTH_ORCHARD,
+    },
+    note::commitment::ExtractedNoteCommitment,
+    primitives::sinsemilla::{i2lebsp_k, HashDomain},
+};
+use pasta_curves::{arithmetic::FieldExt, pallas};
+
+use ff::{Field, PrimeFieldBits};
 use rand::RngCore;
+use std::iter;
 
 /// The root of an Orchard commitment tree.
-#[derive(Clone, Debug)]
+#[derive(Eq, PartialEq, Clone, Debug)]
 pub struct Anchor(pub [u8; 32]);
 
+impl From<pallas::Base> for Anchor {
+    fn from(anchor_field: pallas::Base) -> Anchor {
+        Anchor(anchor_field.to_bytes())
+    }
+}
+
 #[derive(Debug)]
-pub struct MerklePath;
+pub struct MerklePath {
+    position: u32,
+    auth_path: [pallas::Base; MERKLE_DEPTH_ORCHARD],
+}
 
 impl MerklePath {
     /// Generates a dummy Merkle path for use in dummy spent notes.
-    pub(crate) fn dummy(_rng: &mut impl RngCore) -> Self {
-        MerklePath
+    pub(crate) fn dummy(rng: &mut impl RngCore) -> Self {
+        fn dummy_inner(rng: &mut impl RngCore, _idx: usize) -> pallas::Base {
+            pallas::Base::random(rng)
+        }
+
+        MerklePath {
+            position: rng.next_u32(),
+            auth_path: gen_const_array(rng, dummy_inner),
+        }
+    }
+
+    /// <https://zips.z.cash/protocol/protocol.pdf#orchardmerklecrh>
+    /// The layer with 2^n nodes is called "layer n":
+    ///      - leaves are at layer MERKLE_DEPTH_ORCHARD = 32;
+    ///      - the root is at layer 0.
+    /// `l_star` is MERKLE_DEPTH_ORCHARD - layer - 1.
```

Comment:
This is what I called "level" in https://github.com/ebfull/incrementalmerkletree/pull/1